core/src/apps/common/keychain.py

```diff
@@ -0,0 +1,190 @@
+from storage import device
+from trezor import wire
+from trezor.crypto import bip32
+
+from . import HARDENED, paths
+from .seed import Slip21Node, get_seed
+
+if False:
+    from typing import (
+        Any,
+        Awaitable,
+        Callable,
+        Dict,
+        List,
+        Optional,
+        Sequence,
+        TypeVar,
+    )
+    from typing_extensions import Protocol
+
+    from protobuf import MessageType
+
+    T = TypeVar("T")
+
+    class NodeProtocol(Protocol[paths.PathType]):
+        def derive_path(self, path: paths.PathType) -> None:
+            ...
+
+        def clone(self: T) -> T:
+            ...
+
+        def __del__(self) -> None:
+            ...
+
+    NodeType = TypeVar("NodeType", bound=NodeProtocol)
+
+    MsgIn = TypeVar("MsgIn", bound=MessageType)
+    MsgOut = TypeVar("MsgOut", bound=MessageType)
+
+    Handler = Callable[[wire.Context, MsgIn], Awaitable[MsgOut]]
+    HandlerWithKeychain = Callable[[wire.Context, MsgIn, "Keychain"], Awaitable[MsgOut]]
+
+    class Deletable(Protocol):
+        def __del__(self) -> None:
+            ...
+
+
+FORBIDDEN_KEY_PATH = wire.DataError("Forbidden key path")
+
+
+class LRUCache:
+    def __init__(self, size: int) -> None:
+        self.size = size
+        self.cache_keys = []  # type: List[Any]
+        self.cache = {}  # type: Dict[Any, Deletable]
+
+    def insert(self, key: Any, value: Deletable) -> None:
+        if key in self.cache_keys:
+            self.cache_keys.remove(key)
+        self.cache_keys.insert(0, key)
+        self.cache[key] = value
+
+        if len(self.cache_keys) > self.size:
+            dropped_key = self.cache_keys.pop()
+            self.cache[dropped_key].__del__()
+            del self.cache[dropped_key]
+
+    def get(self, key: Any) -> Any:
+        if key not in self.cache:
+            return None
+
+        self.cache_keys.remove(key)
+        self.cache_keys.insert(0, key)
+        return self.cache[key]
+
+    def __del__(self) -> None:
+        for value in self.cache.values():
+            value.__del__()
+        self.cache.clear()
+        self.cache_keys.clear()
+        del self.cache
+
+
+class Keychain:
+    def __init__(
+        self,
+        seed: bytes,
+        curve: str,
+        namespaces: Sequence[paths.Bip32Path],
+        slip21_namespaces: Sequence[paths.Slip21Path] = (),
+        restrict: bool = True,
+    ) -> None:
+        # if `restrict` is False, namespaces will not be checked
+        self.seed = seed
+        self.curve = curve
+        self.namespaces = namespaces
+        self.slip21_namespaces = slip21_namespaces
+        self.restrict = restrict
```

Comment:
I am not sure how much sense it makes to set `restrict` per Keychain instance, because right now it's considered a global setting. On the other hand I would like to see the Keychain class moving in a direction, where it will just be holding the derived nodes that it was authorized to hold, i.e. it shouldn't keep a copy of the seed at all.

Reply:
Re `restrict`, it's reasonable that the Keychain would look at the setting directly; I'm honestly not sure why I did it this way originally. I'll remove it.

Reply:
7289d84 and 02fcf99